### DISCUSSION TITLE
How to enable CORS when using the development server combined with a second local server (local host) ?

### DISCUSSION THREAD

**Discussion by serge-hulne** [mint-lang/mint#558] [Answered] [answered]
How to enable CORS when using the development server 

`mint start`  // runs on localhost:3000

combined with a second local server (local host) // running on e.g. localhost:8080

**Comment by gdotdesign** [comment, marked as answer]
Hey @serge-hulne 

CORS is done on the server side, you need to set up the appropriate response headers for all requests:
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#the_http_response_headers

**Reply by serge-hulne** [reply]
Thanks , but how do you do this when using the built-in (“mint start”) development server?

More precisely, it would be advantageous in some circumstances to be able to use the developments server in combination with a second server which provides data, for instance.

In that scheme, the development server would be used to develop the font end interface (using Mint), whereas the second server would be used to provide the external stream of data which the front Interface has to process.

**Reply by gdotdesign** [reply]
The development server just serves the HTML, CSS and compiled JavaScript to run the application in the browser. It's the browser who makes requests to the second server (in your setup) which triggers CORS, and thus the second server needs to add the response headers I linked earlier to its responses.

Let me know if it's not the case, in which case I'm probably misunderstanding your setup.

**Reply by serge-hulne** [reply]
I think it is the other way around. Here is a snippet from my Main.mint and from my Kemal server (Crystal), respectively:

The test server (mint start) runs on its default port (3000) and the local data server runs on port 3000, respectively:

Main.mint
========
```
store PersonStore {
  state status : Status = Status::Initial
  state object : String = "valeur initiale"

  fun load : Promise(Never, Void) {
    sequence {
      next { status = Status::Loading }
      response =
        "http://127.0.0.1:2000/data/"
        |> Http.get()
        |> Http.send()

      object =
        response.body
        result = Data(object)
        Debug.log(result)
        next { status = Status::Ok(result) }
    }
    catch Http.ErrorResponse => error {
      next { status = Status::Error("Something went wrong with the request.") }
    }  
  }
}
``` 

server.cr
=======

```
require "kemal"


# ============
# Settings
# ============

IP = "127.0.0.1"
PORT = 2000

Kemal.config.port = (ENV["PORT"]? || PORT).to_i
Kemal.config.host_binding = ENV["HOST_BINDING"]? || "#{IP}"
# Kemal.config.env = "production"

# =======
# Actions
# =======

static_headers do |response, filepath, filestat|
    response.headers.add("Access-Control-Allow-Origin", "*")
end

# Kemal.config.public_folder = "../../dist"

get "/data/" do
  "Hello World!"
end


Kemal.run
```
Error message received in the dev tools (console) in the browser:
================================================
```
Access to XMLHttpRequest at 'http://127.0.0.1:2000/data/' from origin 'http://127.0.0.1:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Reply by gdotdesign** [reply]
That's correct, but you only added the response headers to the static files, not the route you defined. This works (I tested it):

```
get "/data/" do |env|
  env.response.headers.add("Access-Control-Allow-Origin", "*")
  "Hello World!"
end
```

**Reply by gdotdesign** [reply]
Here is a discussion about this specific case `static_headers` https://github.com/kemalcr/kemal/issues/420#issuecomment-343451413

**Reply by serge-hulne** [reply]
Thank you very much!
